### PR TITLE
Refactor definition-based extraction for multiple resource

### LINKED ISSUE #670: Refactor definition-based extraction for multiple resource and create extraction context
**Is your feature request related to a problem? Please describe.**
The extraction code for multiple resources needs to be refactored

**Describe the solution you'd like**
Refactor the definition based extraction functions so that they are not extension functions of Base any more. Instead, a extraction context should be passed down during extraction."

We need to create something like `data class ExtractionContext` and pass it down during the definition-based extraction process.

It should probably include the current target resource, as well as anything else that has been created to be added to the final bundle.

**Describe alternatives you've considered**
Current approach does work, but it's difficult to read and understand and lacks the level of abstraction we need for this code.

**Additional context**
NA

**Would you like to work on the issue?**
NA

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #670 

**Description**
Clear and concise code change description. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Code health

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by santosh-pingle** [comment]
@jingtang10 

- I have removed base extension type of the functions.
- I tried to create extension context data class which have current resource, bundle, questionnaire and response item
`data class ExtractionContext(
  var currentResource: Base,
  val bundle: Bundle,
  var questionnaireItem: Questionnaire.QuestionnaireItemComponent,
  var questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent
)`
- But not sure how to use `Map<ResourceType, List<Resource>>` which was mentioned in the issue, as bundle is passed in the extraction context to collect the extracted resource.

Created first draft, can you please review once and provide your input on it.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 1]
```diff
@@ -620,3 +651,10 @@ private fun Base.asExpectedType(): Type {
     else -> this as Type
   }
 }
+
+data class ExtractionContext(
+  val currentResource: Base,
+  val bundle: Bundle,
+  val questionnaireItem: Questionnaire.QuestionnaireItemComponent,
+  val questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:659]: I think the extraction context should include a list of resources to be extracted, and a list of resources that have been extracted.

so every time you come across a new extraction context extension, you add the new resource to the list of resources to be extracted, and then you recursively visit the child questions, and then you move that resource to the list of resources that have been extracted.

At the end of your traversal of the questionnaire, you'll end up with the first list being empty, and the second list containing all the resources that have been extracted.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
Thanks @santosh-pingle and sorry for the delay.

Please read through http://build.fhir.org/ig/HL7/sdc/extraction.html#definition-based-extraction if you haven't. I think the details in this section will inform the design decisions in this PR.

**Comment by santosh-pingle** [comment]
Working on review comments

**Comment by santosh-pingle** [comment]
It's still in progress.
I am trying to make changes as per review comments.
thanks!

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 1 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt]: Code is refactored as mush as possible as per review comments.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 2]
```diff
@@ -637,3 +667,13 @@ private fun Base.asExpectedType(): Type {
...
 }
+
+/** A data class which holds extraction data to extract and update the resource. */
+internal data class ExtractionContext(
+  var currentResource: Base,
+  val extractedResourceList: MutableList<Resource>,
+  var questionnaireItem: Questionnaire.QuestionnaireItemComponent,
+  var questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent,
+  var questionnaireItemList: List<Questionnaire.QuestionnaireItemComponent>,
+  var questionnaireResponseItemList: List<QuestionnaireResponse.QuestionnaireResponseItemComponent>
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:678]: i don't think these should be in the context.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 3]
```diff
@@ -224,13 +228,10 @@ object ResourceMapper {
...
    * the corresponding questions in [questionnaireItemList]. This method handles nested fields.
    */
-  private suspend fun Base.extractFields(
-    bundle: Bundle,
-    questionnaireItemList: List<Questionnaire.QuestionnaireItemComponent>,
-    questionnaireResponseItemList: List<QuestionnaireResponse.QuestionnaireResponseItemComponent>
-  ) {
-    val questionnaireItemListIterator = questionnaireItemList.iterator()
-    val questionnaireResponseItemListIterator = questionnaireResponseItemList.iterator()
+  private suspend fun extractFields(extractionContext: ExtractionContext) {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:231]: i think the questionnaire items and questionnaire response items still should be what you use in the recursion. this function (and the other functions) are basically doing a tree-traversal. and the extraction context is what is passed down in the traversal to carry the extraction progress and result.

so please add those back to the param list in this function.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by santosh-pingle** [comment]
looking into the review comments.

**Comment by jingtang10** [comment]
I took the liberty of rewriting some part of the resource mapper in #1014, which is merged after speaking with @santosh-pingle and @aditya-07. It covers a little more refactoring than in this PR - notably changing the code structure for definition based extraction.

Thanks again @santosh-pingle for this work.

**Pull request closed without merging**